File: backend/golang/internal/engine/workflow_step_executors.go
package engine

import (
	"context"
	"fmt"
	"log"
	"time"

	"rocket-backend/internal/metadata"
	"rocket-backend/internal/store"
)

// StepResult represents the outcome of executing a workflow step.
type StepResult struct {
	Paused   bool
	NextGoto string
}

// StepExecutorContext provides dependencies that step executors need.
type StepExecutorContext struct {
	ActionExecutors map[string]ActionExecutor
	Evaluator       ExpressionEvaluator
	Registry        *metadata.Registry
}

// StepExecutor handles execution of a single workflow step type.
type StepExecutor interface {
	Execute(ctx context.Context, q store.Querier, ectx *StepExecutorContext, instance *metadata.WorkflowInstance, step *metadata.WorkflowStep) (*StepResult, error)
}

// ActionStepExecutor runs all actions in an action step sequentially.
type ActionStepExecutor struct{}

func (e *ActionStepExecutor) Execute(ctx context.Context, q store.Querier, ectx *StepExecutorContext,
	instance *metadata.WorkflowInstance, step *metadata.WorkflowStep) (*StepResult, error) {

	for _, action := range step.Actions {
		executor, ok := ectx.ActionExecutors[action.Type]
		if !ok {
			log.Printf("WARN: unknown workflow action type: %s", action.Type)
			continue
		}
		if err := executor.Execute(ctx, q, ectx.Registry, instance, &action); err != nil {
			return nil, fmt.Errorf("action %s: %w", action.Type, err)
		}
	}

	instance.History = append(instance.History, metadata.WorkflowHistoryEntry{
		Step:   step.ID,
		Status: "completed",
		At:     time.Now().UTC().Format(time.RFC3339),
	})

	next := ""
	if step.Then != nil {
		next = step.Then.Goto
	}
	return &StepResult{Paused: false, NextGoto: next}, nil
}

// ConditionStepExecutor evaluates a boolean expression and branches.
type ConditionStepExecutor struct{}

func (e *ConditionStepExecutor) Execute(_ context.Context, _ store.Querier, ectx *StepExecutorContext,
	instance *metadata.WorkflowInstance, step *metadata.WorkflowStep) (*StepResult, error) {

	if step.Expression == "" {
		return nil, fmt.Errorf("condition step %s has no expression", step.ID)
	}

	env := map[string]any{
		"context": instance.Context,
	}

	isTrue, err := ectx.Evaluator.EvaluateBool(step.Expression, env)
	if err != nil {
		return nil, err
	}

	status := "on_false"
	next := ""
	if isTrue {
		status = "on_true"
		if step.OnTrue != nil {
			next = step.OnTrue.Goto
		}
	} else {
		if step.OnFalse != nil {
			next = step.OnFalse.Goto
		}
	}

	instance.History = append(instance.History, metadata.WorkflowHistoryEntry{
		Step:   step.ID,
		Status: status,
		At:     time.Now().UTC().Format(time.RFC3339),
	})

	return &StepResult{Paused: false, NextGoto: next}, nil
}

// ApprovalStepExecutor pauses the workflow and optionally sets a deadline.
type ApprovalStepExecutor struct{}

func (e *ApprovalStepExecutor) Execute(_ context.Context, _ store.Querier, _ *StepExecutorContext,
	instance *metadata.WorkflowInstance, step *metadata.WorkflowStep) (*StepResult, error) {

	if step.Timeout != "" {
		duration, err := time.ParseDuration(step.Timeout)
		if err == nil {
			deadline := time.Now().UTC().Add(duration).Format(time.RFC3339)
			instance.CurrentStepDeadline = &deadline
		}
	}

	return &StepResult{Paused: true, NextGoto: ""}, nil
}

// DefaultStepExecutors returns the built-in set of step executors.
func DefaultStepExecutors() map[string]StepExecutor {
	return map[string]StepExecutor{
		"action":    &ActionStepExecutor{},
		"condition": &ConditionStepExecutor{},
		"approval":  &ApprovalStepExecutor{},
	}
}

File: backend/expressjs/src/engine/workflow-step-executors.ts
import type { Queryable } from "../store/postgres.js";
import type { WorkflowInstance, WorkflowStep } from "../metadata/workflow.js";
import type { ActionExecutor } from "./workflow-action-executors.js";
import type { ExpressionEvaluator } from "./workflow-expression.js";

/**
 * StepResult represents the outcome of executing a workflow step.
 */
export interface StepResult {
  paused: boolean;
  nextGoto: string;
}

/**
 * StepExecutorContext provides dependencies that step executors need.
 */
export interface StepExecutorContext {
  actionExecutors: Map<string, ActionExecutor>;
  evaluator: ExpressionEvaluator;
}

/**
 * StepExecutor handles execution of a single workflow step type.
 * Implementations exist for action, condition, and approval steps.
 */
export interface StepExecutor {
  execute(
    q: Queryable,
    ctx: StepExecutorContext,
    instance: WorkflowInstance,
    step: WorkflowStep,
  ): Promise<StepResult>;
}

/**
 * ActionStepExecutor runs all actions in an action step sequentially,
 * then advances to the next step.
 */
export class ActionStepExecutor implements StepExecutor {
  async execute(
    q: Queryable,
    ctx: StepExecutorContext,
    instance: WorkflowInstance,
    step: WorkflowStep,
  ): Promise<StepResult> {
    for (const action of step.actions ?? []) {
      const executor = ctx.actionExecutors.get(action.type);
      if (executor) {
        await executor.execute(q, instance, action);
      } else {
        console.warn(`WARN: unknown workflow action type: ${action.type}`);
      }
    }

    instance.history.push({
      step: step.id,
      status: "completed",
      at: new Date().toISOString(),
    });

    const next = step.then?.goto ?? "";
    return { paused: false, nextGoto: next };
  }
}

/**
 * ConditionStepExecutor evaluates a boolean expression and branches
 * to on_true or on_false paths.
 */
export class ConditionStepExecutor implements StepExecutor {
  async execute(
    _q: Queryable,
    ctx: StepExecutorContext,
    instance: WorkflowInstance,
    step: WorkflowStep,
  ): Promise<StepResult> {
    if (!step.expression) {
      throw new Error(`condition step ${step.id} has no expression`);
    }

    const env = { context: instance.context };
    let result: boolean;
    try {
      result = ctx.evaluator.evaluateBool(step.expression, env);
    } catch (err: any) {
      throw new Error(`evaluate condition: ${err.message ?? err}`);
    }

    let status: string;
    let next = "";
    if (result) {
      status = "on_true";
      next = step.on_true?.goto ?? "";
    } else {
      status = "on_false";
      next = step.on_false?.goto ?? "";
    }

    instance.history.push({
      step: step.id,
      status,
      at: new Date().toISOString(),
    });

    return { paused: false, nextGoto: next };
  }
}

/**
 * ApprovalStepExecutor pauses the workflow and optionally sets a deadline
 * for timeout handling.
 */
export class ApprovalStepExecutor implements StepExecutor {
  async execute(
    _q: Queryable,
    _ctx: StepExecutorContext,
    instance: WorkflowInstance,
    step: WorkflowStep,
  ): Promise<StepResult> {
    if (step.timeout) {
      const ms = parseDuration(step.timeout);
      if (ms > 0) {
        const deadline = new Date(Date.now() + ms).toISOString();
        instance.current_step_deadline = deadline;
      }
    }

    return { paused: true, nextGoto: "" };
  }
}

/**
 * Creates the default set of step executors.
 */
export function createDefaultStepExecutors(): Map<string, StepExecutor> {
  const executors = new Map<string, StepExecutor>();
  executors.set("action", new ActionStepExecutor());
  executors.set("condition", new ConditionStepExecutor());
  executors.set("approval", new ApprovalStepExecutor());
  return executors;
}

/**
 * Parse duration strings like "72h", "48h", "30m" to milliseconds.
 */
function parseDuration(s: string): number {
  const match = s.match(/^(\d+)(h|m|s)$/);
  if (!match) return 0;
  const num = parseInt(match[1], 10);
  switch (match[2]) {
    case "h": return num * 60 * 60 * 1000;
    case "m": return num * 60 * 1000;
    case "s": return num * 1000;
    default: return 0;
  }
}
